weather_data_analyzer.py

```diff
@@ -0,0 +1,179 @@
+import csv
+import sys
+from typing import List
+
+from weather_record import WeatherRecord
+from custom_types import (AverageTempHumidReportType,
+                          HighLowTempHumidityReportType)
+
+
+class WeatherDataAnalyzer:
+    """
+    Receives the file paths and store their record in the list. Then perform
+    different calculations on that list
+    """
+
+    def __init__(self, filepaths) -> None:
+
+        self.file_records: List[WeatherRecord] = []
+
+        self.file_data_reader_and_add_to_file_records(filepaths)
+
+    def file_data_reader_and_add_to_file_records(self, filepaths):
+        """
+        Read the data of all files given in filepaths and append it to the list
+
+        :param filepaths:
+        """
+
+        for filepath in filepaths:
+            with open(filepath, 'r') as file:
+                records_list = csv.DictReader(file)
+
+                for row in records_list:
+                    record = WeatherRecord(row)
+                    self.file_records.append(record)
+
+    def calc_low_and_high_temp_and_humid(self):
+
+        if len(self.file_records) == 0:
+            return None
+
+        return self.get_low_high_temp_and_max_humidity()
+
+    def get_low_high_temp_and_max_humidity(self):
+        """
+        From the file_records list, it returns the max of high temperature, min
+        of low temperature and max of high humidity
+
+        :return: HighLowTempHumidityReportType
+        """
+
+        valid_high_temp_records = self.remove_none_values_of_max_temp()
+        valid_low_temp_records = self.remove_none_values_of_min_temp()
+        valid_high_humidity_records = self.remove_none_values_of_max_humidity()
+
+        max_high_temp = max(
+            valid_high_temp_records, key=lambda x: int(x.max_temp)
+        )
+        min_low_temp = min(
+            valid_low_temp_records, key=lambda x: int(x.min_temp)
+        )
+        max_high_humidity = max(
+            valid_high_humidity_records, key=lambda x: int(x.max_humidity)
+        )
+
+        return HighLowTempHumidityReportType(
+            highest_temperature=max_high_temp,
+            lowest_temperature=min_low_temp,
+            high_humidity=max_high_humidity
+        )
+
+    def remove_none_values_of_max_temp(self):
+        """
```

Comment:
these multiple methods can be converted to single, pass the relevant data and it should remove the None values from it.

Reply:
I think They can't be converted into single loop. Our data is like : 
```python
| max_temp | min_temp | max_humidity | mean_humidity |
|----------|----------|--------------|---------------|
| None     | 3        | None         | 20            |
| 38       | None     | None         | None          |
| None     | None     | 100          | None          |
| 36       | 2        | 95           | 50            |

```

```python
ans = []

for rec in table_records:
     if rec.max_temp and rec.min_temp and rec.max_humidity and rec.mean_humidity:
           ans.append(rec)


```

If i use a single loop with single array `ans`  to remove None values, then I will get the single row left from table (the last row in which all values of each column are not None) `| 36       | 2        | 95           | 50            |`.

To calculate the highest `max_temp`, I need all valid values of `max_temp` column i.e. [38, 36]. The above single loop technique is removing the `max_temp 38 - 2nd row` because other columns values in this row are None.

The other option is that use single loop with four different arrays:
```python
max_temp = [], min_temp = [], max_humid = [], mean_humid = []

for rec in table_records:
     if rec.max_temp:
          max_temp.append(rec)

    if rec.min_temp:
          min_temp.append(rec)

...........
```
It will mess up the code and four arrays will return (in tuple/dict). so that's why i have used four different functions. Let me know, if the above four array with single loop is good in a way, then i will go for it.